theano/scan_module/scan_perform.pyx

```diff
@@ -366,9 +378,26 @@ def perform(
 
         # Check which of the pre-allocated outputs (if applicable) have
         # been reused by the inner function
-        for j in range(len_output_storage):
-            output_reused[j] = (old_output_storage[j] is
-                                output_storage[j].storage[0])
+        for idx in range(len_output_storage):
+            # If the storage map does not contain the same object, then
+            # the pre-allocated output has not been reused
+            new_var = output_storage[idx].storage[0]
+            if old_output_storage[idx] is new_var:
+
+                # The pre-allocated output is only considered as having
+                # been reused if it still points to the same data as it
+                # did before the execution of the inner function
+                if old_output_data[idx] is None:
+                    output_reused[idx] = False
+                else:
+                    if hasattr(new_var, 'gpudata'):
+                        output_reused[idx] = (new_var.gpudata ==
+                                              old_output_data[idx])
```

Comment:
As a bonus, this will work correctly for the new backend too!  Not that I reuse old python objects though.